File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long number = get_long("Number: ");
    int number_lenght = 0;
    int digit;
    int first_two_digits;
    int first_digit;
    int odd_sum = 0;
    int even_sum = 0;
    while (number != 0)
    {
        number_lenght++;
        digit = number % 10;
        if (number_lenght % 2 == 0)
        {
            digit *= 2;
            do
            {
                even_sum += digit % 10;
                digit /= 10;
            }
            while (digit != 0);
        }
        else
        {
            odd_sum += digit;
        }
        if (number < 100 && number >= 10)
        {
            first_two_digits = number;
        }
        else if (number < 10)
        {
            first_digit = number;
        }
        number /= 10;
    }
    if ((even_sum + odd_sum) % 10 == 0)
    {
        if (number_lenght == 15 && (first_two_digits == 34 || first_two_digits == 37))
        {
            printf("AMEX\n");
        }
        else if (number_lenght == 16 && (first_two_digits == 51 || first_two_digits == 52 
                                         || first_two_digits == 53 || first_two_digits == 54 
                                         || first_two_digits == 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((number_lenght == 13 || number_lenght == 16) && first_digit == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: pset6/pset6/credit/credit.py
from sys import exit

# imput credit card number from user
card_number = input("Number: ")

# Luhn’s Algorithm
# total sum
sum = 0

# loop through every other digit, starting with second-to-last digit
for digit in card_number[-2::-2]:
    digit = int(digit) * 2
    if digit > 9:
        digit = str(digit)
        sum += int(digit[0]) + int(digit[1])
    else:
        sum += digit

# add remaining digits
for digit in card_number[::-2]:
    sum += int(digit)

# check if card number is valid
if sum % 10 != 0:
    print("INVALID")
    exit(0)

# credit card number length
length = len(card_number)

# check type of credit card
if length == 15 and card_number[:2] in ["37", "34"]:
    print("AMEX")
elif length == 16 and card_number[:2] in ["51", "52", "53", "54", "55"]:
    print("MASTERCARD")
elif (length == 13 or length == 16) and card_number[0] == "4":
    print("VISA")
else:
    print("INVALID")
